contrib/babelfishpg_common/src/spatialtypes.c

```
@@ -67,134 +242,975 @@ load_functions()
+    {
+        /* Check if geometry is empty */
+        UpdateFunctionCallInfo(fcinfo_local, 1, geom_datum);
+        is_empty_datum = st_isempty_p(fcinfo_local);
+        is_empty = DatumGetBool(is_empty_datum);
+
+        if (!is_empty) 
+        {
+            /* Validate latitude for non-empty geography points */
+            validate_latitude(geom_datum);
+        }
+    }
+
+    return geom_datum;
+}
+
+/* This function converts WKT (Well-Known Text) input to a geometry object. */
+Datum 
+charTogeom(PG_FUNCTION_ARGS) 
+{
+    return char_to_geo_common(PG_GETARG_TEXT_PP(0), false);
+}
+
+/* This function converts WKT (Well-Known Text) input to a geography object. */
+Datum 
+charTogeog(PG_FUNCTION_ARGS) 
+{
+    return char_to_geo_common(PG_GETARG_TEXT_PP(0), true);
+}
+
+/* Helper function implementations for bytea to geometry/geography conversions */
+static void 
+validate_input_length(const bytea *input, const char *type_name) 
+{
+    if (VARSIZE_ANY_EXHDR(input) < 22) {
         ereport(ERROR,
-            (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-            errmsg("%s is not supported", geometry_name)));
-    else
+                (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+                 errmsg("Invalid %s", type_name)));
+    }
+}
+
+static GeometryData* 
+initialize_geometry_data(bytea *input) 
+{
+    GeometryData *geom_data = palloc(sizeof(GeometryData));
+    
+    geom_data->input = input;
+    geom_data->input_data = (uint8 *)VARDATA_ANY(input);
+    geom_data->input_len = VARSIZE_ANY_EXHDR(input);
+    
+    /* Extract SRID from first 4 bytes (little-endian) */
+    geom_data->srid = (geom_data->input_data[3] << 24) | 
+                      (geom_data->input_data[2] << 16) | 
+                      (geom_data->input_data[1] << 8) | 
+                       geom_data->input_data[0];
+    
+    /* Extract geometry type */
+    geom_data->geom_type = (geom_data->input_data[4] << 8) | 
+                            geom_data->input_data[5];
```

This is incorrect! Consider the following 2 scenarios:
1. Point Empty byte sequence -> here geom type is the final byte and not 5th and 6th bytes.
2. What if 5th and 6th bytes are `0104`(also valid type for linestring, etc.) or any other valid type, with current implementation they would considered as `POINT EMPTY` but rather should throw an error.

For eg:
```
1> select CAST(CAST(0x000000000104000000000000000001000000ffffffffffffffff02 as geometry) As varchar(MAX))
2> go
```
```
1> select CAST(CAST(0x000000000104000000000000084000000000000069400000000000001040 as geometry) As varchar(MAX))
2> go
```
Please correct the logic and we also need to add more test combinations around Casting to/from varbinary to/from geometry/geography.

---

EMPTY_COORD[] = {
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x01, 0x00, 0x00, 0x00, 0xff, 0xff, 0xff, 0xff,
    0xff, 0xff, 0xff, 0xff
};
 We will match the whole binary format of TSQL (EMPTY_COORD[] ) with input to verify whether it is really an empty point since PostGIS and TSQL have totally different representations of Empty geometries. Let's validate empty geometry format:
 Input length should be sizeof(EMPTY_COORD) + 7 bytes 
 (6 bytes header + 20 bytes coordinates + 1 byte type)
 Compare input data (skipping 6 byte header) with EMPTY_COORD pattern
 Throw error if either condition fails